lib/puma/server.rb

```diff
@@ -476,19 +480,19 @@ def process_client(client)
 
             requests += 1
 
-            # As an optimization, try to read the next request from the
-            # socket for a short time before returning to the reactor.
-            fast_check = @status == :run
+            Thread.pass
 
-            # Always pass the client back to the reactor after a reasonable
-            # number of inline requests if there are other requests pending.
-            fast_check = false if requests >= @max_fast_inline &&
-              @thread_pool.backlog > 0
+            client.reset
 
-            next_request_ready = with_force_shutdown(client) do
-              client.reset(fast_check)
+            # This indicates that the socket has pipelined (multiple)
+            # requests on it, so process them
+            next_request_ready = if client.has_buffer
+              with_force_shutdown(client) { client.fast_try_to_finish }
+            else
+              nil
             end
 
+            # Send keep-alive connections back to the reactor
             unless next_request_ready
```

Comment:
Optional style comment: This would read better to me as `if !pipeline_request_ready`. I checked and rubocop wouldn't mind that syntax.

Or alternatively get rid of the negation:

```
if pipeline_request_ready
  continue
else 
  # ...
```

I like that better actually as it makes it more obvious we're in a loop that will execute again if that condition is true.